File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
import sys
input = sys.stdin.readline


def solution(cap, n, deliveries, pickups):
    answer = 0

    while deliveries or pickups:

        while deliveries and deliveries[-1] == 0:
            deliveries.pop()
        while pickups and pickups[-1] == 0:
            pickups.pop()

        if len(deliveries) >= len(pickups):
            answer += len(deliveries)
        else:
            answer += len(pickups)

        temp_cap = cap
        while temp_cap > 0 and deliveries:
            if temp_cap >= deliveries[-1]:
                temp_cap -= deliveries[-1]
                deliveries.pop()
            else:
                deliveries[-1] -= temp_cap
                break

        temp_cap = cap
        while temp_cap > 0 and pickups:
            if temp_cap >= pickups[-1]:
                temp_cap -= pickups[-1]
                pickups.pop()
            else:
                pickups[-1] -= temp_cap
                break

    return answer * 2

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        int d = deliveries.length - 1;
        int p = pickups.length - 1;

        while (d != -1 && deliveries[d] == 0) {
            d -= 1;
        }
        while (p != -1 && pickups[p] == 0) {
            p -= 1;
        }

        long answer = 0;
        while (d != -1 || p != -1) {
            int v = Math.max(d, p) + 1;
            answer += v * 2L;

            d = getLastIndex(cap, deliveries, d);
            p = getLastIndex(cap, pickups, p);
        }

        return answer;
    }

    private int getLastIndex(int cap, int[] datas, int idx) {
        int back = cap;
        while (idx != -1) {
            if (datas[idx] <= back) {
                back -= datas[idx];
                idx -= 1;
            } else {
                datas[idx] -= back;
                break;
            }
        }
        return idx;
    }
}